bindings/haskell/src/layers.rs

```diff
@@ -0,0 +1,74 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::ffi::c_char;
+
+use ::opendal as od;
+
+#[repr(C)]
+pub struct Layer {
+    tag: LayerTag,
+    data: LayerData,
+}
+
+#[allow(dead_code)]
+#[repr(C)]
+pub enum LayerTag {
```

Comment:
I don't want to expose layers to other language bindings this way. In fact, hashkell need to implement bindings in hashkell instead, like I proposed in https://github.com/apache/incubator-opendal/discussions/1960.

Reply:
We need to implement a HaskellLayer which implements `opendal::Layer`, and then, HaskellLayer should expose as a haskell API which allows user to implement in haskell.

Reply:
Because we expose all operations as typeclass in haskell, users can complete functions similar to `layer` according to writing their own `OpMonad`. Maybe what we can do is to expose some layers writing by rust.

Reply:
> Maybe what we can do is to expose some layers writing by rust.

I think we can enable `retry` layer by default. And we don't need to provide API for users to configure layers in rust core for now.